keps/sig-api-machinery/20200421-static-admission-webhook-configuration.md

```diff
@@ -0,0 +1,149 @@
+objects. Any policy enforcement is not in place until these webhook
+configurations are created, thereby registering the webhook. This creates a gap
+in enforcement spanning from when the kube-apiserver is started to until the
+webhook configuration objects are created and picked up by the dynamic admission
+controller. Another gap is the inability of the cluster administrator to protect
+against deletion of these webhook configuration objects as
+MutatingWebhookConfiguration and ValidatingWebhookConfiguration objects are not
+subject to webhook admission policy.
+
+This KEP aims to address these issues.  
+
+### Goals
+
+- A robust admission webhook registration process where there is no period of
+  time between the kube-apiserver coming up and the registration of an admission
+  webhook.
+
+- The registration of webhooks registered through this process should be
+  protected from alteration by API requests made by cluster users.
+
+- The new registration process should also have a highly privileged means of
+  being able to alter webhook registrations.
+ 
+## Proposal
+
+In short, this proposal is about augmenting `AdmissionConfig`'s plugin
+`configuration` [specification](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers)
+to include a path to a configuration file containing one of the following:
+- `admissionregistration.k8s.io/v1.ValidatingWebhookConfigurationList` 
+- `admissionregistration.k8s.io/v1.MutatingWebhookConfigurationList` 
+- v1.List with items of admissionregistration.k8s.io/v1.ValidatingWebhookConfiguration 
+- v1.List  with items of admissionregistration.k8s.io/v1.MutatingWebhookConfiguration.
+This configuration is loaded  as static webhooks in api servers. These
+webhooks are called by validating and mutating admission plugins along with
+dynamically loaded webhooks for relevant admission requests. This also means
+that these webhooks are not API visible objects and hence cannot be modified by
+an API user.
+
+### New AdmissionConfig schema
+
+A `staticConfigFile` field is added to `configuration` field of a plugin object in
+`AdmissionConfig`.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: ValidatingAdmissionWebhook
+  configuration:
+    ....
+    staticConfigFile: "<path-to-static-config-file>"
+- name: MutatingAdmissionWebhook
+  configuration:
+    ....
+    staticConfigFile: "<path-to-static-config-file>"
+```
+
+### Reconfiguring static webhook configuration
+
+The static configuration file is watched so that the static webhook
+configuration can be changed by editing the contents of the file.
+
+## Drawbacks
+
+- Reduced visibility for users wanting to list all active admission webhooks.
+  This KEP has similar visibility characteristics as compiled in admission
+  controllers.
+- Webhooks hosted in-cluster can still be disabled by cluster users with enough
+  privilege. 
```

Comment:
this doesn't seem to be a drawback with this proposal, just a note that would inform decisions about where/how to host the backing webhook server

Reply:
Makes sense. Removed